Esp32s2 changes

tom-van opened the pull request:
HW SPI did not work on esp32-s2:
```
MicroPython v1.15-74-gfb631644e on 2021-05-04; ESP32S2 module with ESP32S2
Type "help()" for more information.
>>> esp.osdebug(0,esp.LOG_DEBUG)
>>> from machine import SPI
>>> spi1=SPI(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: SPI(1) doesn't exist
>>> spi2=SPI(2)
E (128705) spi: spi_bus_initialize(632): invalid dma channel
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: invalid configuration
```
The problem is caused by different SPI_HOST aliases for CONFIG_IDF_TARGET_ESP32S2
in esp-idf/components/hal/include/hal/spi_types.h:
```
#define FSPI_HOST   SPI2_HOST
#define HSPI_HOST   SPI3_HOST
```
There is no VSPI_HOST so SPI(1) does not exist.
HSPI_HOST moved from SPI2 to SPI3. That's why SPI(2) tried wrong DMA channel.

MicroPython v1.15 is no longer compatible with IDF 3.x so we can avoid [HVF]SPI_HOST
aliases and use SPI[123]_HOST directly. Also from IDF 4.3 we can use SPI_DMA_CH_AUTO
instead of selecting proper channel number.

The commit "esp32-s2: set default SPI pins for GENERIC_S2 board" sets proper default pins
for SPI(1) and SPI(2). The question is if we should not change the default pin definitions
MICROPY_HW_SPI1_SCK... based on CONFIG_IDF_TARGET_ESP32S2

The last commit "esp32-s2: make GPIO 26 usable if SPIRAM is not configured"
is not related to SPI - just adds GPIO 26 if not used as SPIRAM CS

tom-van committed:
ports/esp32: S2,S3: Make GPIO 26 usable if SPIRAM is not configured.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

dpgeorge commented:
Thanks for the contribution!

The DMA issue was fixed in 13e6e0d7f5dc830639fd21e2b1380c6cc1996cab (although maybe not as neat as is done here with SPI_DMA_CH_AUTO...).

The pin issue was fixed in 5093d49fae5b59d6440d34b4c05f2950f40a6b9f

------

The change left in this PR is the change to `GENERIC_S2` board:

> The question is if we should not change the default pin definitions
MICROPY_HW_SPI1_SCK... based on CONFIG_IDF_TARGET_ESP32S2

Yes I think that would be a good idea.  If you could update this PR to do that, that'd be great.

tom-van committed:
ports/esp32: Set proper default SPI(id=1) pins on S2,S3 and C3.

Use IO_MUX pins as defined by ESP IDF in soc/esp32*/include/soc/spi_pins.h
Alternatively use now deprecated HSPI_IOMUX_PIN_NUM_xxx
(or FSPI_IOMUX_PIN_NUM_xxx for ESP32S2) for compatibility with IDF 4.2
and older.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van committed:
ports/esp32: Set proper default SPI(id=2) pins on S2 and S3.

Use IO_MUX pins as defined by ESP IDF in soc/esp32/include/soc/spi_pins.h

ESP32S2 and S3 does not have IO_MUX pins for SPI3, GPIO matrix is always
used. Choose suitable defaults for S2 and S3.

ESP32C3 does not have SPI3 at all. Do not define pin mappings for it.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van commented:
@dpgeorge 

> The pin issue was fixed in [5093d49](https://github.com/micropython/micropython/commit/5093d49fae5b59d6440d34b4c05f2950f40a6b9f)

Since mentioned commit up to the current git master GPIO 26 is configured as unusable for both S2 and S3
regardless SPIRAM is used or not. So we would profit from
https://github.com/micropython/micropython/pull/7291/commits/c648c041d5814b91181cacfb109c70cd240a97dd
I updated the commit msg of the old change to succeed in workflow tests.

> > The question is if we should not change the default pin definitions
> > MICROPY_HW_SPI1_SCK... based on CONFIG_IDF_TARGET_ESP32S2
> 
> Yes I think that would be a good idea. If you could update this PR to do that, that'd be great.

Done

dpgeorge commented on ports/esp32/machine_pin.c:
```diff
@@ -177,7 +177,11 @@ STATIC const machine_pin_obj_t machine_pin_obj[] = {
     {{NULL}, -1}, // 24 not a pin
     {{NULL}, -1}, // 25 not a pin
-    {{NULL}, -1}, // 26 FLASH/PSRAM
+    #if CONFIG_ESP32_SPIRAM_SUPPORT
```
This constant is ESP32 specific, ie won't work on ESP32S2/S3.  I think `CONFIG_SPIRAM` can be used as a generic value, but I didn't test it.

tom-van replied:
@dpgeorge thanks and sorry!

Fixed as you proposed.

Not sure if CONFIG_SPIRAM is really generic (it isn't defined in [boards/sdkconfig.spiram](https://github.com/micropython/micropython/blob/master/ports/esp32/boards/sdkconfig.spiram)) but I checked it is defined in 
[boards/sdkconfig.spiram_sx](https://github.com/micropython/micropython/blob/master/ports/esp32/boards/sdkconfig.spiram_sx) what is enough for this particular purpose. To be sure I poisoned one of #if CONFIG_SPIRAM branches and compiled for BOARD=GENERIC_S3_SPIRAM and BOARD=GENERIC_S2

dpgeorge commented on ports/esp32/machine_pin.c:
```diff
@@ -618,7 +622,11 @@ STATIC const machine_pin_irq_obj_t machine_pin_irq_object[] = {
     {{NULL}, -1}, // 24 not a pin
     {{NULL}, -1}, // 25 not a pin
-    {{NULL}, -1}, // 26 FLASH/PSRAM
+    #if CONFIG_ESP32_SPIRAM_SUPPORT
```
same as above

tom-van replied:
Fixed

tom-van commented:
I found a substantial problem in SPI pins defaults:
My old commit 0a9fd1e957a824bbc7f95d32073c7e1edc61d1b9 assigned SPI(1) to IDF SPI2 and SPI(2) to IDF SPI3 on all devices.
On the other hand https://github.com/micropython/micropython/commit/13e6e0d7f5dc830639fd21e2b1380c6cc1996cab
uses HSPI  for SPI(1). It means that on ESP32 we get IDF SPI2 and on ESP32Sx we get IDF SPI3. Similarly SPI(2) assigns to IDF SPI3 (VSPI) on ESP32 and to SPI2 (FSPI) on ESP32Sx. I can adjust defaults for this but the #if  #endif mess will be horrible.

tom-van commented:
The latest commits are adjusted for reordered SPI on S2, S3

tom-van committed:
ports/esp32: Fix SPI default pins reordering on ESP32S2 and S3.

The index of machine_hw_spi_obj and machine_hw_spi_default_pins
arrays is assigned to 0 for ARG_id==HSPI_HOST and 1 for another SPI.
On ESP32S2 and S3 HSPI_HOST=2 so the first set (idx=0) of default
pins is used for SPI(id=2) aka HSPI/SPI3 and the second set (idx=1)
for SPI(id=1) aka FSPI/SPI2.
This makes misleading mess in MICROPY_HW_SPIxxxx definitions
and it is also in contradiction to the comments around the definitions.

Change the test of ARG_id to fix the order of machine_hw_spi_default_pins.

This change might require adjusting MICROPY_HW_SPIxxxx definitions
in mpconfigboard.h of S2/S3 based boards.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van committed:
ports/esp32/boards: Remove SPI pin defaults from GENERIC S2/S3 boards.

Default SPI pins are now correctly assigned by machine_hw_spi.c
even for S2 and S3.
mpconfigboard.h files define defaults with flipped SPI(1) and SPI(2)
to workaround a bug in machine_hw_spi.c - the bug is fixed.

Remove defaults.

Signed-off-by: Tomas Vanek <vanekt@fbl.cz>

tom-van commented:
@dpgeorge , @UnexpectedMaker 
Assigning of SPI default pins now finally works even on S2, S3 and should work also on C3 (not tested).
I checked remaining SPI pin definitions in mpconfigboard.h of S2, S3 boards and they look correct - and now apply to SPI(1) as originally expected and not to SPI(2) due to the bug.
The only exception is
https://github.com/micropython/micropython/blob/a7fa18c203a241f670f12ab507aa8b349fcd45a1/ports/esp32/boards/LOLIN_S2_PICO/mpconfigboard.h#L11-L12
where SCK and MISO are flipped in comparison to https://www.wemos.cc/en/latest/s2/s2_pico.html
But this is out of my interest.

dpgeorge commented:
Thanks for the effort on this, making sure it is all correct and fixing the bug with id=HSPI.  It looks good now.

Merged in f9733705a907a9e9c3b3e6ab613cc1d7f63b51a6 through 9aa151e3f355dea6e14d77053f40691365e60d96
